lib/harmony/harmonytask/task_type_handler.go

```diff
@@ -95,12 +97,33 @@ top:
 		return false
 	}
 
+	var location string
+	releaseStorage := func() {
+	}
+	if h.TaskTypeDetails.Cost.Storage != nil {
+		if c := h.TaskTypeDetails.Cost.Storage.Claim; c != nil {
+			if err = c(int(*tID)); err != nil {
+				log.Infow("did not accept task", "task_id", strconv.Itoa(int(*tID)), "reason", "storage claim failed", "name", h.Name, "error", err)
+				return false
+			}
+			h.LocationMap.Store(*tID, location)
+			releaseStorage = func() {
+				if err := h.TaskTypeDetails.Cost.Storage.MarkComplete(); err != nil {
+					log.Errorw("Could not release storage", "error", err)
+				}
+				h.LocationMap.Delete(*tID)
+			}
+		}
+	}
+
 	// if recovering we don't need to try to claim anything because those tasks are already claimed by us
 	if from != workSourceRecover {
 		// 4. Can we claim the work for our hostname?
 		ct, err := h.TaskEngine.db.Exec(h.TaskEngine.ctx, "UPDATE harmony_task SET owner_id=$1 WHERE id=$2 AND owner_id IS NULL", h.TaskEngine.ownerID, *tID)
 		if err != nil {
 			log.Error(err)
+
+			releaseStorage()
 			return false
 		}
 		if ct == 0 {
```

Comment:
We also must call releaseStorage if we didn't accept the taks because it was already taken